tempodb/backend/azure/azure_helpers.go

```diff
@@ -59,88 +54,112 @@ func GetContainerURL(ctx context.Context, cfg *Config, hedge bool) (blob.Contain
+	var client *azblob.Client
+	accountKey := getStorageAccountKey(cfg)
+
+	switch {
+	case cfg.UseFederatedToken:
+		credential, err := azidentity.NewWorkloadIdentityCredential(&azidentity.WorkloadIdentityCredentialOptions{})
 		if err != nil {
-			return blob.ContainerURL{}, err
+			return container.Client{}, err
 		}
 
-		p = blob.NewPipeline(credential, opts)
-	} else {
-		credential, err := getOAuthToken(cfg)
+		client, err = azblob.NewClient(u.String(), credential, &opts)
+
 		if err != nil {
-			return blob.ContainerURL{}, err
+			return container.Client{}, err
 		}
+	case cfg.UseManagedIdentity:
+		var id azidentity.ManagedIDKind
 
-		p = blob.NewPipeline(*credential, opts)
-	}
+		if cfg.UserAssignedID != "" {
+			id = azidentity.ClientID(cfg.UserAssignedID)
+		}
 
-	accountName := getStorageAccountName(cfg)
-	u, err := url.Parse(fmt.Sprintf("https://%s.%s", accountName, cfg.Endpoint))
+		// azidentity.NewManagedIdentityCredential defaults to a system-assigned identity.
+		// We only set options.ID if we want a user-assigned identity.
+		// See azidentity.ManagedIdentityCredential.
+		credential, err := azidentity.NewManagedIdentityCredential(&azidentity.ManagedIdentityCredentialOptions{
+			ID: id,
+		})
+		if err != nil {
+			return container.Client{}, err
+		}
 
-	// If the endpoint doesn't start with blob.core we can assume Azurite is being used
-	// So the endpoint should follow Azurite URL style
-	// https://learn.microsoft.com/en-us/rest/api/storageservices/get-blob#emulated-storage-service-uri
-	if !strings.HasPrefix(cfg.Endpoint, "blob.") {
-		u, err = url.Parse(fmt.Sprintf("http://%s/%s", cfg.Endpoint, accountName))
-	}
+		client, err = azblob.NewClient(u.String(), credential, &opts)
 
-	if err != nil {
-		return blob.ContainerURL{}, err
-	}
+		if err != nil {
+			return container.Client{}, err
+		}
+	case accountName != "" && accountKey != "":
+		credential, err := azblob.NewSharedKeyCredential(accountName, getStorageAccountKey(cfg))
+		if err != nil {
+			return container.Client{}, err
+		}
 
-	service := blob.NewServiceURL(*u, p)
+		client, err = azblob.NewClientWithSharedKeyCredential(u.String(), credential, &opts)
 
-	return service.NewContainerURL(cfg.ContainerName), nil
-}
+		if err != nil {
+			return container.Client{}, err
+		}
+	// If no authentication mechanism has been explicitly specified, assume workload identity.
+	default:
+		credential, err := azidentity.NewWorkloadIdentityCredential(&azidentity.WorkloadIdentityCredentialOptions{})
+		if err != nil {
+			return container.Client{}, err
+		}
 
-func GetContainer(ctx context.Context, conf *Config, hedge bool) (blob.ContainerURL, error) {
-	c, err := GetContainerURL(ctx, conf, hedge)
-	if err != nil {
-		return blob.ContainerURL{}, err
+		client, err = azblob.NewClient(u.String(), credential, &opts)
+
+		if err != nil {
+			return container.Client{}, err
+		}
 	}
-	return c, nil
+
+	return *client.ServiceClient().NewContainerClient(cfg.ContainerName), nil
```

Comment:
it seems the sdk prefers to return a pointer to container.Client? any reason not to follow suit in `getContainerClient`. seems like it would be simpler

Reply:
> any reason not to follow suit in `getContainerClient`

The main reason is that the old SDK implementation did not use pointers, but it seems that was because the old SDK didn't return pointers.

> seems like it would be simpler

Agreed.

Action taken: we now return pointers for SDK clients.